python/ray/_private/import_thread.py

```diff
@@ -75,22 +70,27 @@ def _run(self):
                 if msg is None:
                     self.threads_stopped.wait(timeout=0.01)
                     continue
-
                 if msg["type"] == "subscribe":
                     continue
-                assert msg["data"] == b"rpush"
-                num_imports = self.redis_client.llen("Exports")
-                assert num_imports >= num_imported
-                for i in range(num_imported, num_imports):
-                    num_imported += 1
-                    key = self.redis_client.lindex("Exports", i)
-                    self._process_key(key)
+                self._do_importing()
         except (OSError, redis.exceptions.ConnectionError, grpc.RpcError) as e:
             logger.error(f"ImportThread: {e}")
         finally:
             # Close the pubsub client to avoid leaking file descriptors.
             import_pubsub_client.close()
 
+    def _do_importing(self):
+        while True:
```

Comment:
while loop seems unnecessary?

Reply:
Yes, it is. This is trying to import everything after num_imported. Think about this, for msg 0-20, we might lose 10-20 from pubsub, and then we get 21, so we'll load 11-21. I think we probably can do better here later, like when subscribe, there is a timeout, and it'll be able to load the missing message even there is an error.